lib/Dialects/LinalgExt/Transforms/InParallelToSequentialFor.cpp

```diff
@@ -0,0 +1,115 @@
+//
+//===---------------------------------------------------------------------===//
+
+#include <mlir/IR/BlockAndValueMapping.h>
+#include <mlir/IR/BuiltinOps.h>
+
+#include "Dialects/LinalgExt/LinalgExtOps.h"
+#include "Dialects/LinalgExt/PassDetail.h"
+#include "Dialects/LinalgExt/Passes.h"
+#include "Transforms/Utils.h"
+#include "llvm/ADT/STLExtras.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Identifier.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::linalg_ext;
+
+namespace {
+
+struct InParallelOpToSCFRewriter
+    : public OpRewritePattern<linalg_ext::InParallelOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg_ext::InParallelOp inParallelOp,
+                                PatternRewriter &rewriter) const override {
+    // Construct the loop bounds based on the canonical arithmetic progression.
+    Location loc = inParallelOp.getLoc();
+    Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Value numThreads = inParallelOp.num_threads();
+
+    // Construct the op without a body builder: we need to clone the ops in the
+    // body explicitly after having access to the new bbArgs.
+    // As a consequence, `ensureTerminator` is not called and the body has no
+    // terminator.
+    PerformConcurrentlyOp performConcurrentlyOp = inParallelOp.getTerminator();
+    scf::ForOp forOp = rewriter.create<scf::ForOp>(
+        loc, zero, numThreads, one, performConcurrentlyOp.yieldedValues());
+
+    rewriter.setInsertionPointToStart(forOp.getBody());
+
+    BlockAndValueMapping bvm;
+    bvm.map(performConcurrentlyOp.yieldedValues(), forOp.getRegionIterArgs());
+
+    // Move the body while replacing the threadId by the forOp iv.
+    SmallVector<Value> bbArgsTranslated{forOp.getInductionVar()};
+    rewriter.mergeBlocks(&inParallelOp.region().front(), forOp.getBody(),
+                         bbArgsTranslated);
+    // Create sequential insertSlice ops.
+    SmallVector<Value> toYield;
+    rewriter.setInsertionPoint(performConcurrentlyOp);
+    for (ParallelInsertSliceOp op : performConcurrentlyOp.yieldingOps()) {
+      toYield.push_back(rewriter.createOrFold<tensor::InsertSliceOp>(
+          loc, op.source(), bvm.lookup(op.dest()), op.offsets(), op.sizes(),
+          op.strides()));
+    }
+
+    // performConcurrentlyOp.yieldedValues come from above, not from bbArgs.
+    // There is no rewriter method to make mergeBlocks update non-bbArgs.
+    // Need to manually clone + bvm all uses that are now nested under forOp.
```

Comment:
I actually think that this is incorrect. If a loop iteration uses a "yielded value" then it should see the tensor value from before the `in_parallel` block runs, not a partially updated one.

Reply:
I am not sure I follow, the scf.for does an incremental partial update at each sequential iteration.
This bufferizes to partial updates to the memory region at each iteration.
How would you write the following in the sequential case ?

```
func @static_tile(%arg0: index, %arg1: tensor<?xf32>, %arg2: tensor<?xf32>) -> ... {
  %2 = scf.for %arg3 = %c0 to %1 step %c1 iter_args(%arg4 = %arg2) -> (tensor<?xf32>) {
    %6 = tensor.extract_slice %arg4[%3] [%5] [%c1] : tensor<?xf32> to tensor<?xf32>
    %7 = tensor.extract_slice %arg1[%3] [%5] [1] : tensor<?xf32> to tensor<?xf32>
    %8 = linalg.generic {indexing_maps = [#map4, #map4], iterator_types = ["parallel"]} ins(%7 : tensor<?xf32>) outs(%6 : tensor<?xf32>) {...}
    %9 = tensor.insert_slice %8 into %arg4[%3] [%5] [%c1] : tensor<?xf32> into tensor<?xf32>
    scf.yield %9 : tensor<?xf32>
  }
}
```

Reply:
Adding a scary comment as per our offline chat.